### PR TITLE
external_signer: improve fingerprint matching logic (stop on first match)

### PR DESCRIPTION
**Pull request by theStack**
The fingerprint matching logic in `ExternalSigner::SignTransaction` currently always iterates all inputs of a PSBT, even after a match has already been found. I guess the reason for that is not that it was not thought of, but rather the fact that breaking out of a nested loop is simply not possible (at least not without adding ugly constructs like gotos or extra state variables).
This PR fixes this by using `std::any_of` from C++'s standard library, see http://www.cplusplus.com/reference/algorithm/any_of/

### CODE REVIEW

**Comment by meshcollider** [comment]
Concept ACK

`breaking out of a nested loop is simply not possible (at least not without adding ugly constructs like gotos or extra state variables)`

You could do this though:

```diff
-for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
+for (unsigned int i = 0; !match && i < psbtx.inputs.size(); ++i) {
    const PSBTInput& input = psbtx.inputs[i];
    for (const auto& entry : input.hd_keypaths) {
-        if (m_fingerprint == strprintf("%08x", ReadBE32(entry.second.fingerprint))) match = true;
+        if (m_fingerprint == strprintf("%08x", ReadBE32(entry.second.fingerprint))) { 
+            match = true;
+            break;
+        }
    }
}
```

**Review by lsilva01 — APPROVED** [review]
Code Review ACK https://github.com/bitcoin/bitcoin/pull/22789/commits/d047ed729f1d4732d23324fc76849f3c657cdbe4

**Comment by Sjors** [comment]
utACK d047ed7

I don't think there was a deep reason I didn't do this.

**Review by Sjors — APPROVED** [review]
utACK d047ed7

I don't think there was a deep reason I didn't do this.

**Review by Zero-1729 — APPROVED** [review]
crACK d047ed729f1d4732d23324fc76849f3c657cdbe4

**Review by mjdietzx — COMMENTED** [review]
Code review ACK d047ed729f1d4732d23324fc76849f3c657cdbe4

File: src/external_signer.cpp [hunk 1]
```diff
@@ -75,15 +76,14 @@ bool ExternalSigner::SignTransaction(PartiallySignedTransaction& psbtx, std::str
...
-    for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
-        const PSBTInput& input = psbtx.inputs[i];
+    auto matches_signer_fingerprint = [&](const PSBTInput& input) {
         for (const auto& entry : input.hd_keypaths) {
-            if (m_fingerprint == strprintf("%08x", ReadBE32(entry.second.fingerprint))) match = true;
+            if (m_fingerprint == strprintf("%08x", ReadBE32(entry.second.fingerprint))) return true;
         }
-    }
+        return false;
+    };
```
Comment (mjdietzx) [inline src/external_signer.cpp:84]: I'm curious, could you get rid of the other for loop as well with the same approach? Would that be cleaner? Something like:

```c++
return std::any_of(input.hd_keypaths.begin(), input.hd_keypaths.end(), [](auto& entry){ return m_fingerprint == strprintf("%08x", ReadBE32(entry.second.fingerprint)); });
```

**Comment by hebasto** [comment]
Concept ACK.

**Review by hebasto — APPROVED** [review]
ACK d047ed729f1d4732d23324fc76849f3c657cdbe4, I have reviewed the code and it looks OK, I agree it can be merged.

**Pull request merged**